judge/ratings.py

```diff
@@ -28,94 +67,77 @@ def tie_ranker(iterable, key=attrgetter('points')):
-        PerfAs = old_rating[i] + CF * (APerf - EPerf)
-        Weight = 1.0 / (1 - (0.42 / (times_rated[i] + 1) + 0.18)) - 1.0
-        if old_rating[i] > 2500:
-            Weight *= 0.8
-        elif old_rating[i] >= 2000:
-            Weight *= 0.9
-
-        Cap = 150.0 + 1500.0 / (times_rated[i] + 2)
-
-        new_rating[i] = (old_rating[i] + Weight * PerfAs) / (1.0 + Weight)
-
-        if abs(old_rating[i] - new_rating[i]) > Cap:
-            if old_rating[i] < new_rating[i]:
-                new_rating[i] = old_rating[i] + Cap
-            else:
-                new_rating[i] = old_rating[i] - Cap
-
-        if times_rated[i] == 0:
-            new_volatility[i] = 385
-        else:
-            new_volatility[i] = math.sqrt(((new_rating[i] - old_rating[i]) ** 2) / Weight +
-                                          (old_volatility[i] ** 2) / (Weight + 1))
-
-        if is_disqualified[i]:
-            # DQed users can manipulate TopCoder ratings to get higher volatility in order to increase their rating
-            # later on, prohibit this by ensuring their volatility never increases in this situation
-            new_volatility[i] = min(new_volatility[i], old_volatility[i])
-
-    # try to keep the sum of ratings constant
-    adjust = float(sum(old_rating) - sum(new_rating)) / N
-    new_rating = list(map(adjust.__add__, new_rating))
-    # inflate a little if we have to so people who placed first don't lose rating
-    best_rank = min(actual_rank)
-    for i in range(N):
-        if abs(actual_rank[i] - best_rank) <= 1e-3 and new_rating[i] < old_rating[i] + 1:
-            new_rating[i] = old_rating[i] + 1
-    return list(map(int, map(round, new_rating))), list(map(int, map(round, new_volatility)))
+        # Calculate performance.
+        solve_idx(0)
+        solve_idx(n - 1)
+        divconq(0, n - 1)
+
+        # Calculate mean.
+        for i, r in enumerate(ranking):
+            tanh_terms = []
+            w_prev = 1.
+            w_sum = 0.
+            for j, h in enumerate(historical_p[i]):
+                gamma2 = (VAR_PER_CONTEST if j > 0 else 0)
+                h_var = get_var(times_ranked[i] + 1 - j, cache)
+                k = h_var / (h_var + gamma2)
+                w = w_prev * k**2
+                # Future optimization: If j is around 20, then w < 1e-3 and it is possible to break early.
+                tanh_terms.append((h, BETA2 ** .5 * TANH_C, w))
+                w_prev = w
+                w_sum += w / BETA2
+            w0 = 1. / get_var(times_ranked[i] + 1, cache) - w_sum
+            p0 = eval_tanhs(tanh_terms[1:], old_mean[i]) / w0 + old_mean[i]
+            new_mean[i] = solve(tanh_terms, w0 * p0, lin_factor=w0)
+
+    # Display a slightly lower rating to incentivize participation.
+    # As times_ranked increases, new_rating converges to new_mean.
+    new_rating = [round(m - (get_var(t + 1, cache) ** .5 - VAR_LIM ** .5)) for m, t in zip(new_mean, times_ranked)]
```

Comment:
Is there a reason you are using `** .5` other than `sqrt` (`from math import sqrt`)?

```console
$ python3 -m timeit -s 'a = 1337' 'a ** .5'
5000000 loops, best of 5: 82.8 nsec per loop
$ python3 -m timeit -s 'a = 1337; from math import sqrt' 'sqrt(a)'
10000000 loops, best of 5: 26.8 nsec per loop
```

Also since `VAR_LIM` is a constant, might as well define it as one?

Reply:
Not exactly sure how timeit works, but I see (consistent) different results. Perhaps in your instance, `timeit` imports `sqrt` every iteration?
```py
>>> timeit.timeit("sqrt(1337)", setup="from math import sqrt", number=1000000)
0.053775883017806336
>>> timeit.timeit("1337**0.5", number=1000000)
0.010119114012923092
```

Reply:
Disregard, variable assignment is important.
```py
>>> timeit.timeit("a=1337; sqrt(a)", setup="from math import sqrt", number=1000000)
0.03669788900879212
>>> timeit.timeit("a=1337; a**0.5", number=1000000)
0.07756782200885937
```
```py
>>> timeit.timeit("a=1337; a=sqrt(a)", setup="from math import sqrt", number=1000000)
0.06084884298616089
>>> timeit.timeit("a=1337; a=a**0.5", number=1000000)
0.098016113974154
```